daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/PartialTransactionSpec.scala

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package speedy
+
+import com.daml.lf.language.{Util => AstUtil}
+import com.daml.lf.speedy.PartialTransaction._
+import com.daml.lf.transaction.{Node, TransactionVersion}
+import com.daml.lf.value.Value
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+class PartialTransactionSpec extends AnyWordSpec with Matchers {
+
+  private[this] val transactionSeed = crypto.Hash.hashPrivateKey("PartialTransactionSpec")
+  private[this] val templateId = data.Ref.Identifier.assertFromString("pkg:Mod:Template")
+  private[this] val choiceId = data.Ref.Name.assertFromString("choice")
+  private[this] val cid = Value.ContractId.V1(crypto.Hash.hashPrivateKey("My contract"))
+
+  private[this] val initialState = PartialTransaction.initial(
+    _ => TransactionVersion.maxVersion,
+    data.Time.Timestamp.Epoch,
+    InitialSeeding.TransactionSeed(transactionSeed),
+  )
+
+  private[this] def contractIdsInOrder(ptx: PartialTransaction): Seq[Value.ContractId] = {
+    ptx.finish match {
+      case CompleteTransaction(tx) =>
+        tx.fold(Vector.empty[Value.ContractId]) {
+          case (acc, (_, create: Node.NodeCreate[Value.ContractId])) => acc :+ create.coid
+          case (acc, _) => acc
```

Comment:
I assume `fold` recurses over every node rather than just root nodes? I couldn’t find the implementation from a quick look.

Reply:
```
  /** Traverses the transaction tree in pre-order traversal (i.e. exercise nodes are traversed before their children)
    *
    * Takes constant stack space. Crashes if the transaction is not well formed (see `isWellFormed`)
    */
  final def fold[A](z: A)(f: (A, (Nid, Node.GenNode[Nid, Cid])) => A): A
```